### DISCUSSION TITLE
Proposal: Support allowing primary constructors to be marked as read-only

### DISCUSSION THREAD

**Discussion by DavidArno** [dotnet/csharplang#7254] [Language Ideas] [answered]
Updated with Cyrus's correction re init-only properties:

As it currently stands the only way to achieve apparent read-only primary constructor (PC) parameters is via read-only fields or init-only properties that effectively hide the parameters. For record classes and readonly struct records, set-only properties are synthesised that replace the parameter reference outside of construction/initialization. For readonly struct classes, any mutation of the parameter symbol outside of construction/initialization results in an error. In all of these cases though, the parameters themselves remain mutable and can be changes during construction/initialization.

PC parameters are scoped to the whole type, so have a significantly greater scope than method parameters. As such, this proposal is focused solely on PC parameters as supporting the ability to specify that PC parameters should be read-only is an order of magnitude more important than for methods. The comparable variable type with the same scope as PC parameters are fields. Because the scope of fields is so great, these have supported being marked `readonly` for many years. As the scope of PC parameters is the same, there seems a compelling case for allowing them to be marked `readonly` too.

It is suggested therefore that `readonly` be made a valid prefix to PC paramters. A parameter decorated with `readonly` could behave in one of two ways:

1. It behaves like a `readonly` field: it can still be mutated during construction/initialization. but modifying it thereafter is an error. This creates consistency with the behaviour of fields, but may be surprising behaviour to many.
2. It cannot be mutated even during construction/initialization. Whilst inconsistent with fields, this likely fits in more with what developers would expect of read-only PC parameters.

**Comment by CyrusNajmabadi** [comment]
> For record classes and readonly struct records, set-only properties are synthesised

Just as a minor clarification.  The properties are `init-only` for a record-class.

**Comment by CyrusNajmabadi** [comment]
>  As the scope of PC parameters is the same, there seems a compelling case for allowing them to be marked readonly too.

Generally speaking, i agree with this as well.  However, i would not limit this to just PC parameters as there would be similar value in being able to state this at least for method-parameters (esp. given how large methods can be, and how useful it can be to know that a method parameter is not mutated).  Supporting lambda-parameters is more optional for me, but is possibly on the table.

> It cannot be mutated even during construction/initialization. Whilst inconsistent with fields, this likely fits in more with what developers would expect of read-only PC parameters.

I would go with this interpretation.  It would certainly be what people would expect within a method-body, and i think it would appear to be quite confusing if now it behaved differently for a PC parameter.  

Currently, just for PCs, this doesn't rise to the level of being worthwhile to me.  But i would champion this if we made it work for PCs and any members.

**Comment by CyrusNajmabadi** [comment]
> For record classes and readonly struct records, set-only properties are synthesised that replace the parameter reference outside of construction/initialization. 

I would phrase this more as: "For records, synthesized members with the same name are introduced that are normally found in a closer scope than the parameter when using just the unqualified name of the parameter itself."

As mentioned in the other threads, there is no "replacing" (or "removing", etc.).  There is just scoping.  For example, a parameter with the same name as a field doesn't "replace/remove" the field.  It is simply in a closer scope for that simple name.  But a user can still of course get to the field in a *qualified* fashion (i.e. `this.` that allows them to choose an outer scope).  The same could hold in the future for PC parameters vs. members.  If the parameter was actually 'replaced' then no amount of scope choice could get you to it.  However, if it's more than there's a name in a closer scope, then that still remains a possibility.

**Comment by CyrusNajmabadi** [comment]
> It behaves like a readonly field: it can still be mutated during construction/initialization. but modifying it thereafter is an error. This creates consistency with the behaviour of fields, but may be surprising behaviour to many.

Right, i would not go this route.  If this is just a parameter, it should behave like a parameter.  Now, if we ever add a feature to PCs that would allow you to trivially define a *field* at the PC point, then i would expect that such a field would behave like a field wrt to readonly-ness.

**Comment by CyrusNajmabadi** [comment]
One *minor* concern i have is around syntax options here.  Do you have anything in mind?  The most reasonable seems to be something like:

```c#
public class X(readonly int y) { }
```

This could work, but potentially is ambiguous int the future if we ever want a `declare a field within the PC` feature and we choose to spec it out as `field-modifiers Type Name`.  Now that may or may not be generating a field.  We might need to think about the syntactic implications here, and if we would prefer a different syntax entirely for `declare a field`.

**Reply by HaloFour** [reply]
IMO, I'd have `readonly` only affect the readonly-ness of the parameter/capture and not consider it as a field modifier.  I'd require an accessibility modifier (`private`, `public`, etc.) in order to auto-promote a PC parameter to a field.

**Reply by CyrusNajmabadi** [reply]
That's def a possibility. It's a pity we allow fields without accessibility as it would lead to an unfortunate inconsistency here. But I agree that this is probably ok, and teaching users "no accessibility = parameter" and "accessibility = field" would could be sufficient.

**Reply by HaloFour** [reply]
> It's a pity we allow fields without accessibility as it would lead to an unfortunate inconsistency here

Yeah, I agree.  But that's already the case, given you can declare a field simply with `int y;` already.

I assume that it'd be a bit tight to get `readonly` PC params onto the LDM docket for C# 12 at this point?  You'd want to be pretty sure that anything done wouldn't block future potential proposals.  I think it might be sufficient to allay some of the concerns around PCs, though.

**Reply by CyrusNajmabadi** [reply]
I'll bring it up. But I think it would be highly unlikely for 12. Very full schedule. Maybe for vnext though!

**Comment by DavidArno** [comment]
Linking this to #7353, as it looks like the team are looking seriously at readonly PC parameters, rather than conflating them with (other) method parameters, which is good news.

**Reply by HaloFour** [reply]
It's considered as one option.  Not all team members agree.  Making PC parameters readonly by default and requiring manually declaring/initializing a separate field to allow mutation may also impact other proposals to allow for declaring PC parameters as fields.

**Reply by DavidArno** [reply]
@HaloFour, making the parameters read-only by default is by far the best solution in my view, but when I proposed that before, the limited feedback it got was all negative. It will be interesting to see if it receives more positive feedback if proposed by the team... 😉

**Reply by HaloFour** [reply]
@DavidArno 

> making the parameters read-only by default is by far the best solution in my view, but when I proposed that before, the limited feedback it got was all negative

I don't think the set of problems that arises from doing so goes away or is addressed based on who is proposing the change.  I honestly think it's more about trying to avoid the conversation around `readonly` parameters in general than what should be done about primary constructors.

Like what happens in the following cases:

```cs
public class C1(int x) {
    void M() {
        x = 123; // illegal, x is a readonly capture
    }
}

public class C2(public int x) {
    void M() {
        x = 123; // legal, x is a public field?
    }
}
```

I honestly think that the conversation would loop right back around to wanting/needing a `readonly` modifier at the parameter level, but then does this mean:

```cs
public class C3(readonly int x) {
    // x is promoted to a field with default (private) accessibility?
}
```

IMO, it feels like throwing out the baby with the bathwater, leading to weird warts that will affect future development in this space.

If C# were taking a stronger opinion regarding mutability in general, like Java has, I'd probably have fewer concerns.

**Reply by DavidArno** [reply]
> If C# were taking a stronger opinion regarding mutability in general, like Java has, I'd probably have fewer concerns.

From what you've pointed out to me regarding readonly structs, I'd say that C# has a very strong opinion on mutability: "*immutability is totally 'meh' unless it affects performance*". I get the impression that readonly fields are almost seen as a moment of embarrassing weakness as they seem so contrary to the overall attitude towards immutability by the team 🤣

**Reply by CyrusNajmabadi** [reply]
As mentioned in your other thread, this is distinctly not the case.

**Reply by CyrusNajmabadi** [reply]
> C# has a very strong opinion on mutability: "immutability is totally 'meh' unless it affects performance

You have it backwards. The preference is toward readonlyness.  *However*, being readonly can negatively impact performance (since now you have to copy values to avoid mutation). So you occasionally begrudgingly make something writable because that would be faster. 

This is precisely why we added things like readonly-structs.  We believe in readonly-ness so much, we wanted to ensure you got both the safety, and performance, with an easy modifier on the type that ensures you do the right thing. 

We are also considering this for classes. But it's much less relevant there as the perf concern isn't present.  C# very much likes safety *and* perf. When a feature gets us both, we're more motivated than if it just gives us one.  This is also why we've invested so much in the 'ref' space. Extreme low level perf speedup, while also maintaining high safety.

**Reply by CyrusNajmabadi** [reply]
>  I get the impression that readonly fields are almost seen as a moment of embarrassing weakness

Not a single person in the team, or language designer has put forth that position. Definitely not here, and absolutely no where else afaict.  Please do not make up fictitious strawmen as to our positions.  

Readonly fields are part and parcel of the language, and something we have been evangelizing and recommending since 1.0. Our designs, best practices, and our tooling all point to this. As do all our discussions on this repo.   

If we were to do it over again, 'readonly' would almost certainly be the default for state (not temp storage).  I don't know where you got this impression, but it is likely due to not reading (which you've claimed you don't do in the past) and coming to a bogus conclusion based on biases.

**Reply by HaloFour** [reply]
I think the point is more that regardless of the opinion of the team that the language creates a pit towards mutability due to the fact that where there is an option the default is always to be mutable.  The fact that the language can't put immutable/readonly declarations on equal footing with mutable declarations seems to be the real limiting factor to any discussions, and the language design team does seem to err on the side of consistency in most cases vs. applying stronger opinions.  For example, Java requires that records be readonly, to the extent that you cannot declare additional mutable state within a record.  Java has always required that captures be readonly, initially explicitly but then implicitly when lambda syntax was added.  That's why when Java considered adding a `val` counterpart to `var` they decided that it was unnecessary since it was impossible for a local to ever be shared state.

**Reply by CyrusNajmabadi** [reply]
Right. We're familiar with java there. But we actively don't think that's a good design.  Looking at captures in practice, we see intentional mutation everywhere (same with locals). 

Fundamentally, this just comes down to opinion. Java would prefer to disallow (though their locals are mutable by default), drawing the readonly-line at places we don't agree with. 

For fields though (which David was referring to) we are not at all on the fence about our views there. There is no view of this being an "embarrassing weakness" or any other such nonsense. 

> and the language design team does seem to err on the side of consistency in most cases vs. applying stronger opinions. 

Consistency is a strong opinion though. It's a view that things working more uniformly is actually better for users :-). . Note the other threads where people continually decry inconsistency and how it confuses them. 

Practically, this is why design exists in the guest place. Because it's tradeoffs. Heck we see the *same* people who have in the past argued heavily against features for being inconsistent, then want inconsistency in new features :-). We have to balance that aspect of design against others. And we certainly will be inconsistent if the value is high enough (see collections). 

The question comes down to how important then we think these things are. I know from some users that readonly -ness is just alt the primary factor that affects how they think about code at all. But we have to weigh that against the users who don't think that way, as well as the data we have on things like safety, bug rates, etc to make an informed decision on these things.

**Reply by DavidArno** [reply]
@HaloFour,

That is certainly part of the story. The team have indeed erred on the side of consistency a lot of the time. The frustrating thing there is every time they do that, it becomes harder to deviate from it. If for example the decision had been made back with C# 7 that immutable-first would be the policy for new features, out vars and var patterns could have been immutable. That would then greatly simplify having readonly locals as `let` then solves the other variable cases. If tuples had been made read-only (and they only weren't due to performance considerations) then record structs would logically be immutable too and so the clear expectation would be that PC parameters for non-records would also be readonly. At that point, read-only memebers and types that required `mut` or somesuch keyword to allow mutation could have neatly followed. But none of this occurred and in fact the opposite occurred in many cases with `readonly` being used (squandered even) for purely performance reasons (readonly structs, readonly members, readonly refs).  

I suspect C# will remain a "pit of mutability" language for all time at this stage as trying to fix it now has become too hard and messy. It's a real shame, but there it is. And that's why I think adding readonly parameters to PCs only is likely the way forward here. It fixes the worst case scenario without complicating the language by trying to fix everything else and doesn't introduce an inconsistency by making them read-only by default.

**Reply by HaloFour** [reply]
> Looking at captures in practice, we see intentional mutation everywhere

When I first switched to primarily working in Java I did find the limitation a little obnoxious, but to be honest, I find that I want to mutate such a value exceedingly rare, and the workaround is simple enough (manually wrap in a reference type with a mutable field) that I don't find it remotely onerous.  In the last five years it's probably come up enough times to count on one hand.

I get that the philosophies of the design teams is often not the same.  If anything Java is typically much more reserved.  If anything it's just very interesting to watch the divergent evolution of the two sister languages.  😄 

> Consistency is a strong opinion though. It's a view that things working more uniformly is actually better for users :-). . Note the other threads where people continually decry inconsistency and how it confuses them.

I totally agree.

> The question comes down to how important then we think these things are. I know from some users that readonly -ness is just alt the primary factor that affects how they think about code at all. But we have to weigh that against the users who don't think that way, as well as the data we have on things like safety, bug rates, etc to make an informed decision on these things.

Also agreed.  Given the thought-exercise of redesigning C# I think I'd rather borrow a page from those languages that require a keyword for any declaration, like Scala, that way mutable and immutable would always be on equal footing.

**Reply by CyrusNajmabadi** [reply]
> The team have indeed erred on the side of consistency a lot of the time.

Note that inconsistency is an area where people provide the most feedback about importance. We certainly have the users that want us to change and become inconsistent. But they're by far the vast minority. You're welcome to disagree with that philosophy, but it's not changing any time soon.

**Reply by CyrusNajmabadi** [reply]
> I find that I want to mutate such a value exceedingly rare, 

Right. You can certainly get do without it. However, what we see is *both* that people absolutely do use this and take advantage of it *and* it doesn't bear out that this is a source of problems.  I think we would feel differently if the latter were not true.  Similar to 'null' we'd be much more willing to shake things up for the clear benefits provided. However, the same doesn't hold for assignability of temp state. So this is a case where change would seemingly be fit changes sake.  For appeasement rather than a true net benefit to the ecosystem. 

> as trying to fix it now has become too hard and messy

A core disagreement is whether a "fix" is needed. You are starting from a position that there is a problem. But when we look at the ecosystem, the code and coding patterns out there, the bug rates and all the rest stuff in this arena, we do not actually see a problem.  You seem to be operating under an assumption that we want to change this for locals in c#. I'm stating that we've given it a lot of thought and believe the way it works is actually *good*, and that changing things here, or even making knobs for changing it would be a poor choice.

**Reply by CyrusNajmabadi** [reply]
> manually wrap in a reference type with a mutable field

That brings back memories :-D. The trick we used was often an array with a single element in it. Poor mans closures ftw. 

Fundamentally it will come down to opinion. Java's approach feels like it's intentionally stubbing my toes without providing value (no aid in readability or maintainability for me). So it twas just a tax.  

Interestingly, in all these discussions, people have not put forth information or data that changes that assessment to me for c#.  I get that people like David want this and have a preference one way. But it feels like just a tax without a presented benefit. 

This is why, for example, the calculus hasn't changed at all for locals.  Indeed, even when we brought to ldm last time, there was seemingly no change in thinking there.  The only interest (which is very mixed), relates to mutability of state. There were can at least see pros/cons, and how they can shift around. So it is there we are willing to consider possibly making a change.  We considered, but didn't feel good about one change. We'll discuss another next week.

**Reply by DavidArno** [reply]
> This is why, for example, the calculus hasn't changed at all for locals...

I'll say one thing for this topic: at least it's encouraged a move away from over-using "orthogonal". It looks like "calculus" has become the new word de jour 🤣

**Reply by DavidArno** [reply]
Please do not mark my comments here as "off topic", Cyrus. I started this discussion, therefore if I comment it is by definition on-topic.

**Comment by DavidArno** [comment, marked as answer]
Closing this discussion as it is achieving little beyond encouraging unhelpful comments and behaviour from Cyrus, so isn't achieving anything useful.
